File: contests/lc/2110.NumberofSmoothDescentPeriodsofaStock/solution.go
package numberofsmoothdescentperiodsofastock

// getDescentPeriods returns the number of smooth descent periods.
// A smooth descent run of length k contributes k*(k+1)/2 periods.
func getDescentPeriods(prices []int) int64 {
	var total int64
	runLength := int64(0)

	for i := 0; i < len(prices); i++ {
		if i > 0 && prices[i-1]-prices[i] == 1 {
			runLength++
		} else {
			// End of previous run.
			total += runLength * (runLength + 1) / 2
			runLength = 1
		}
	}

	// Account for the final run.
	total += runLength * (runLength + 1) / 2
	return total
}

File: contests/lc/2110.NumberofSmoothDescentPeriodsofaStock/solution.py
from typing import List
from itertools import groupby
from math import comb
from operator import sub
from itertools import pairwise, starmap

class Solution:
    def getDescentPeriods(self, prices: List[int]) -> int:
        total = 0
        run_length = 0

        for i, price in enumerate(prices):
            if i > 0 and prices[i - 1] - price == 1:
                run_length += 1
            else:
                # Close the previous run and start a new one.
                total += run_length * (run_length + 1) // 2
                run_length = 1

        total += run_length * (run_length + 1) // 2
        return total

class Solution2:
    def getDescentPeriods(self, a: List[int]) -> int:
        return sum(comb(sum(g)+1,2) for d,g in groupby(map(sub,a,a[1:])) if d==1)+len(a)
        return sum(comb(sum(1 for _ in g)+1,2) for d,g in groupby(starmap(sub,pairwise(a))) if d==1)+len(a)